File: python/problem560.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        numsLen = len(nums)
        count = 0
        cumulativeSum = [nums[0]]
        for i in range(1, numsLen):     #计算累加和
            cumulativeSum.append(cumulativeSum[-1] + nums[i])
        cumuSumDict = {k : 1}           #以需要的下一个累加和为key，需要这样的下一个累加和的次数为value
        for cumuSum in cumulativeSum:
            if cumuSum in cumuSumDict:  #当前累加和已在字典中出现，说明当前索引和之前若干个索引之间的区间和为k
                count += cumuSumDict[cumuSum]
            nextCumuSum = cumuSum + k   #想要当前索引和可能的接下来某个索引之间区间和为k，那么其累加和应该在当前基础上+k
            if nextCumuSum in cumuSumDict:
                cumuSumDict[nextCumuSum] += 1
            else:
                cumuSumDict[nextCumuSum] = 1
        return count

File: Java/problem560.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int numsLen = nums.length, count = 0;
        int[] cumulativeSum = new int[numsLen];
        cumulativeSum[0] = nums[0];
        for (int i = 1; i < numsLen; ++i) {
            cumulativeSum[i] = cumulativeSum[i - 1] + nums[i];
        }
        Map<Integer, Integer> cumuSum2count = new HashMap<>();
        cumuSum2count.put(k, 1);
        for (int i : cumulativeSum) {
            if (cumuSum2count.containsKey(i)) {
                count += cumuSum2count.get(i);
            }
            int nextCumuSum = i + k;
            if (cumuSum2count.containsKey(nextCumuSum)) {
                cumuSum2count.put(nextCumuSum, cumuSum2count.get(nextCumuSum) + 1);
            }
            else {
                cumuSum2count.put(nextCumuSum, 1);
            }
        }
        return count;
    }
}
